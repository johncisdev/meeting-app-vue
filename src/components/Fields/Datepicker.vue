<template>
  <v-menu
    ref="menu"
    v-model="menu"
    :close-on-content-click="false"
    offset-y
    max-width="290px"
    min-width="auto"
  >
    <template v-slot:activator="{ on, attrs }">
      <v-text-field
        :rules="rules"
        v-model="inputData"
        label="Meeting Date"
        persistent-hint
        v-bind="attrs"
        @blur="date = parseDate(inputData)"
        v-on="on"
      ></v-text-field>
    </template>
    <v-date-picker
      v-model="inputData"
      no-title
      @change="menu = false"
      @input="$emit('update:inputData', formatDate(inputData))"
    ></v-date-picker>
  </v-menu>
</template>
<script>
export default {
  name: "Datepicker",
  props: ["inputData", "rules"],
  data() {
    return {
      date: new Date(Date.now() - new Date().getTimezoneOffset() * 60000)
        .toISOString()
        .substr(0, 10),
      menu: false,
    };
  },
  // created() {
  //   this.inputData = this.formatDate(this.inputData);
  // },
  computed: {
    computedInputDate() {
      return this.formatDate(this.inputData);
    },
  },
  methods: {
    formatDate(date) {
      if (!date) return null;

      const [year, month, day] = date.split("-");
      return `${month}/${day}/${year}`;
    },
    parseDate(date) {
      if (!date) return null;

      const [month, day, year] = date.split("/");
      return `${year}-${month.padStart(2, "0")}-${day.padStart(2, "0")}`;
    },
  },
};
</script>
