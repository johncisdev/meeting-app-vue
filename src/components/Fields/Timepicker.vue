<template>
  <v-menu
    ref="menu"
    v-model="menu"
    :close-on-content-click="false"
    :nudge-right="40"
    :return-value.sync="inputData"
    transition="scale-transition"
    offset-y
    max-width="290px"
    min-width="290px"
  >
    <template v-slot:activator="{ on, attrs }">
      <v-text-field
        v-model="inputData"
        :label="label"
        readonly
        v-bind="attrs"
        v-on="on"
        :rules="rules"
      ></v-text-field>
    </template>
    <v-time-picker
      v-if="menu"
      v-model="inputData"
      :min="min"
      :max="max"
      :allowed-minutes="allowedMinutes"
      full-width
      @click:minute="$refs.menu.save(inputData)"
      @input="$emit('update:inputData', inputData)"
    ></v-time-picker>
  </v-menu>
</template>
<script>
export default {
  name: "Timepicker",
  props: {
    label: String,
    min: String,
    max: String,
    inputData: String,
    rules: Array,
  },
  data() {
    return {
      menu: false,
    };
  },
  methods: {
    allowedMinutes: (v) => v % 30 === 0,
  },
};
</script>
