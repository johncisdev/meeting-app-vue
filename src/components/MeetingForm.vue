<template>
  <v-row justify="center">
    <v-dialog v-model="isOpen" persistent max-width="600px">
      <v-card>
        <v-card-title>
          <span class="text-h5">{{ id ? "Updated" : "Create" }} Meeting</span>
        </v-card-title>
        <v-card-text>
          <v-container>
            <v-row>
              <v-col cols="12">
                <v-text-field
                  v-model="name"
                  label="Room Name"
                  :rules="rules"
                ></v-text-field>
              </v-col>
              <v-col cols="12">
                <v-combobox
                  v-model="hostName"
                  :items="items"
                  :search-input.sync="search"
                  :rules="rules"
                  hide-selected
                  label="Host Name"
                  persistent-hint
                  small-chips
                >
                  <template v-slot:selection="data">
                    <v-chip
                      :key="JSON.stringify(data.item)"
                      v-bind="data.attrs"
                      :input-value="data.selected"
                      :disabled="data.disabled"
                      @click:close="data.parent.selectItem(data.item)"
                    >
                      <v-avatar
                        class="accent white--text"
                        left
                        v-text="data.item.slice(0, 1).toUpperCase()"
                      ></v-avatar>
                      {{ data.item }}
                    </v-chip>
                  </template>
                  <template v-slot:no-data>
                    <v-list-item v-if="search">
                      <v-list-item-content>
                        <v-list-item-title>
                          Press <kbd>enter</kbd> to add
                          <strong> {{ search }}</strong>
                        </v-list-item-title>
                      </v-list-item-content>
                    </v-list-item>
                  </template>
                </v-combobox>
              </v-col>
              <v-col cols="12">
                <v-combobox
                  v-model="participants"
                  :items="items"
                  :search-input.sync="search"
                  :rules="rules"
                  hide-selected
                  label="Participants"
                  multiple
                  persistent-hint
                  small-chips
                >
                  <template v-slot:selection="data">
                    <v-chip
                      :key="JSON.stringify(data.item)"
                      v-bind="data.attrs"
                      :input-value="data.selected"
                      :disabled="data.disabled"
                      @click:close="data.parent.selectItem(data.item)"
                    >
                      <v-avatar
                        class="accent white--text"
                        left
                        v-text="data.item.slice(0, 1).toUpperCase()"
                      ></v-avatar>
                      {{ data.item }}
                    </v-chip>
                  </template>
                  <template v-slot:no-data>
                    <v-list-item v-if="search">
                      <v-list-item-content>
                        <v-list-item-title>
                          Press <kbd>enter</kbd> to add
                          <strong> {{ search }}</strong>
                        </v-list-item-title>
                      </v-list-item-content>
                    </v-list-item>
                  </template>
                </v-combobox>
              </v-col>
              <v-col cols="12" sm="4">
                <datepicker :inputData.sync="bookingDate" :rules="rules" />
              </v-col>
              <v-col cols="12" sm="4">
                <timepicker
                  :inputData.sync="bookingStartTime"
                  :rules="rules"
                  label="Start Time"
                  min="8:00"
                  max="17:00"
                />
              </v-col>
              <v-col cols="12" sm="4">
                <timepicker
                  :inputData.sync="bookingEndTime"
                  :rules="rules"
                  label="End Time"
                  :min="bookingStartTime"
                  max="17:00"
                />
              </v-col>
              <v-col cols="12" v-show="error">
                <v-snackbar
                  :timeout="-1"
                  :value="true"
                  absolute
                  bottom
                  color="error"
                  left
                  text
                >
                  {{ error }}
                </v-snackbar>
              </v-col>
            </v-row>
          </v-container>
        </v-card-text>
        <v-card-actions>
          <v-spacer></v-spacer>
          <v-btn color="blue darken-1" text @click="openFormDialog">
            Close
          </v-btn>
          <v-btn color="blue darken-1" text @click="submitForm"> Save </v-btn>
        </v-card-actions>
      </v-card>
    </v-dialog>
  </v-row>
</template>

<script>
import Datepicker from "./Fields/Datepicker";
import Timepicker from "./Fields/Timepicker.vue";

export default {
  name: "MeetingForm",
  components: { Datepicker, Timepicker },
  props: {
    "is-open": Boolean,
    "open-form-modal": Function,
    selected: Object,
  },
  data() {
    return {
      rules: [
        (value) => {
          const error = "This Field is required.";
          if (value instanceof Array && value.length == 0) {
            return error;
          }
          return !!value || error;
        },
      ],
      error: "",
    };
  },
  computed: {
    id() {
      return this.selected.id;
    },
    name() {
      return this.selected.name;
    },
    hostName() {
      return this.selected.hostName;
    },
    participants() {
      return this.selected.participants;
    },
    bookingDate() {
      return this.selected.bookingDate;
    },
    bookingStartTime() {
      return this.selected.bookingStartTime;
    },
    bookingEndTime() {
      return this.selected.bookingEndTime;
    },
  },
  //   updated() {
  //     if (Object.keys(this.selected).length) {
  //       Object.keys(this.selected).forEach((key) => {
  //         this[key] = this.selected[key];
  //       });
  //     }
  //   },
  methods: {
    openFormDialog() {
      this.resetForm();
      this.$emit("openFormDialog");
    },
    resetForm() {
      this.name = "";
      this.hostName = "";
      this.participants = [];
      this.bookingDate = "";
      this.bookingStartTime = "";
      this.bookingEndTime = "";
      this.error = "";
    },
    submitForm() {
      const {
        id,
        name,
        hostName,
        participants,
        bookingDate,
        bookingStartTime,
        bookingEndTime,
      } = this;
      const hasSched = this.$store.state.dataList.filter(
        (data) =>
          data.id !== id &&
          data.name === name &&
          data.bookingDate === bookingDate &&
          data.bookingStartTime === bookingStartTime
      );
      if (hasSched.length) {
        this.error = "Unable to save, chosen date and time is not available.";
      } else {
        this.$store.dispatch("saveMeeting", {
          id,
          name,
          hostName,
          participants,
          bookingDate,
          bookingStartTime,
          bookingEndTime,
        });

        this.resetForm();
        this.$emit("openFormDialog");
        this.$emit(
          "showSnackbar",
          `Meeting was successfully ${id ? "updated" : "created"}.`
        );
      }
    },
  },
  watch: {
    bookingStartTime(val) {
      const time = val.split(":");
      const date = new Date(`${this.bookingDate} ${val}`);
      date.setMinutes(date.getMinutes() + 30);
      this.bookingEndTime = `${date.getHours()}:${date.getMinutes()}`;
      return time.join(":");
    },
  },
};
</script>
