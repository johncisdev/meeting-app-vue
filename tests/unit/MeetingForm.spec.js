import { shallowMount, createLocalVue, mount } from "@vue/test-utils";
import Vuex from "vuex";

import MeetingForm from "@/components/MeetingForm.vue";
import Datepicker from "@/components/Fields/Datepicker.vue";
import Timepicker from "@/components/Fields/Timepicker.vue";
import Store from "@/store/index";

describe("MeetingForm.vue test suite", () => {
  const localVue = createLocalVue();
  localVue.use(Vuex);

  let store;
  let propsData;

  beforeEach(() => {
    store = new Vuex.Store(Store);
    propsData = {
      selected: {},
      isOpen: true,
    };
  });

  it("should match with the snapshot", () => {
    const wrapper = mount(MeetingForm, {
      store,
      propsData,
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it("should render the component", () => {
    const wrapper = shallowMount(MeetingForm, {
      store,
      propsData,
    });
    expect(wrapper.exists()).toBe(true);
  });

  it("should be open", () => {
    const wrapper = shallowMount(MeetingForm, {
      store,
      propsData,
    });
    expect(wrapper.props().isOpen).toBe(true);
  });

  it("should render Room Field", () => {
    const wrapper = shallowMount(MeetingForm, {
      store,
      propsData,
    });
    expect(wrapper.find("v-text-field[label='Room Name']").exists()).toBe(true);
  });

  it("should render Host Field", () => {
    const wrapper = shallowMount(MeetingForm, {
      store,
      propsData,
    });
    expect(wrapper.find("v-combobox[label='Host Name']").exists()).toBe(true);
  });

  it("should render Participants Field", () => {
    const wrapper = shallowMount(MeetingForm, {
      store,
      propsData,
    });
    expect(wrapper.find("v-combobox[label='Participants']").exists()).toBe(
      true
    );
  });

  it("should render Meeting Booking Date Field", () => {
    const wrapper = shallowMount(MeetingForm, {
      store,
      propsData,
    });
    expect(wrapper.find(Datepicker).exists()).toBe(true);
  });

  it("should render Meeting Booking Start time Field", () => {
    const wrapper = shallowMount(MeetingForm, {
      store,
      propsData,
    });
    expect(wrapper.find("[label='Start Time']").exists()).toBe(true);
  });

  it("should render Meeting Booking End time Field", () => {
    const wrapper = shallowMount(MeetingForm, {
      store,
      propsData,
    });
    expect(wrapper.find("[label='End Time']").exists()).toBe(true);
  });
});
