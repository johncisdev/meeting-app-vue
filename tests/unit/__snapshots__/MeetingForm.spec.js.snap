// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MeetingForm.vue test suite should match with the snapshot 1`] = `
<v-row
  justify="center"
>
  <v-dialog
    max-width="600px"
    persistent=""
  >
    <v-card>
      <v-card-title>
        <span
          class="text-h5"
        >
          Create Meeting
        </span>
      </v-card-title>
       
      <v-card-text>
        <v-container>
          <v-row>
            <v-col
              cols="12"
            >
              <v-text-field
                label="Room Name"
                rules="value => {
        const error = \\"This Field is required.\\";

        if (value instanceof Array && value.length == 0) {
          return error;
        }

        return !!value || error;
      }"
              />
            </v-col>
             
            <v-col
              cols="12"
            >
              <v-combobox
                hide-selected=""
                label="Host Name"
                persistent-hint=""
                rules="value => {
        const error = \\"This Field is required.\\";

        if (value instanceof Array && value.length == 0) {
          return error;
        }

        return !!value || error;
      }"
                small-chips=""
              />
            </v-col>
             
            <v-col
              cols="12"
            >
              <v-combobox
                hide-selected=""
                label="Participants"
                multiple=""
                persistent-hint=""
                rules="value => {
        const error = \\"This Field is required.\\";

        if (value instanceof Array && value.length == 0) {
          return error;
        }

        return !!value || error;
      }"
                small-chips=""
              />
            </v-col>
             
            <v-col
              cols="12"
              sm="4"
            >
              <v-menu
                max-width="290px"
                min-width="auto"
                offset-y=""
              >
                 
                <v-date-picker
                  no-title=""
                />
              </v-menu>
            </v-col>
             
            <v-col
              cols="12"
              sm="4"
            >
              <v-menu
                max-width="290px"
                min-width="290px"
                nudge-right="40"
                offset-y=""
                transition="scale-transition"
              >
                 
                <!---->
              </v-menu>
            </v-col>
             
            <v-col
              cols="12"
              sm="4"
            >
              <v-menu
                max-width="290px"
                min-width="290px"
                nudge-right="40"
                offset-y=""
                transition="scale-transition"
              >
                 
                <!---->
              </v-menu>
            </v-col>
             
            <v-col
              cols="12"
              style="display: none;"
            >
              <v-snackbar
                absolute=""
                bottom=""
                color="error"
                left=""
                text=""
                timeout="-1"
                value="true"
              >
                
                
              
              </v-snackbar>
            </v-col>
          </v-row>
        </v-container>
      </v-card-text>
       
      <v-card-actions>
        <v-spacer />
         
        <v-btn
          color="blue darken-1"
          text=""
        >
          
          Close
        
        </v-btn>
         
        <v-btn
          color="blue darken-1"
          text=""
        >
           Save 
        </v-btn>
      </v-card-actions>
    </v-card>
  </v-dialog>
</v-row>
`;
